queue/factory.js

```
@@ -0,0 +1,221 @@
+		Body: JSON.stringify(obj),
+		Bucket: bucketName,
+		Key: key,
+		ContentType: 'application/json'
+	});
+
+	/**
+	 * Move object on S3
+	 * @param key
+	 * @param newKey
+	 */
+	const moveObject = (key, newKey) => {
+		const params = {
+			Bucket: bucketName,
+			CopySource: `${bucketName}/${key}`,
+			Key: newKey,
+		};
+
+		return s3Methods.copyObject(params)
+			.then(result => deleteObject(key));
+	};
+
+	/**
+	 * Delete object
+	 * @param key
+	 */
+	const deleteObject = key => s3Methods.deleteObject({
+		Bucket: bucketName,
+		Key: key
+	});
+
+	/**
+	 * Parse JSON from a buffer returned by S3
+	 * @param s3Object
+	 */
+	const parseS3ObjectBody = s3Object => {
+		let result = false;
+
+		// let's not fail the whole program
+		try {
+			result = JSON.parse(s3Object.Body.toString());
+		} catch (e) {
+		}
+
+		return result;
+	};
+
+	/**
+	 * Add objects to queue
+	 * @param arrayOfObjects
+	 * @returns {Array}
+	 */
+	const enqueueItems = function (arrayOfObjects) {
+		return Promise.map(arrayOfObjects, obj => {
+			return putObject(`${queuesItemsPath}/${createObjectname(nameTemplate, obj)}`, obj);
+		}, {concurrency: 5});
+	};
+
+	/**
+	 * Get enqueued objects
+	 */
+	const getItems = co.wrap(function *(limit = 0) {
+		const list = yield readPath(queuesItemsPath, limit);
+
+		return Promise.map(list, item => {
+			return getJSON(item.Key)
+				.then(result => Object.assign({}, result, {queueId: item.Key}));
+		}, {concurrency: 5});
+	});
+
+	/**
+	 * Mark single item as failed
+	 */
+	const markItemFailed = co.wrap(function *(key) {
+		const metaObjectKey = key.replace(queuesItemsPath, metaPath);
+
+		// this will server as default value
+		let metaObject = {attempt: 0};
+
+		/**
+		 * Try and fetch existing meta. Delete it if found,
+		 * since we will save new one.
+		 */
+		try {
+			const storedMeta = yield getJSON(metaObjectKey);
+			yield deleteObject(metaObjectKey);
+			metaObject = Object.assign({}, metaObject, storedMeta);
+		} catch (e) {
+		}
```

At very least let's always add a comment explaining why we have an empty catch.

---

Yes, but since this package will be included in lambdas, I don't wan't lambdas' dependencies to bloat in cases where a whole dependency can be replaced by one simple function.

---

Done